### DISCUSSION TITLE
Adding variants while supporting `group-*` and `peer-*`

### DISCUSSION THREAD

**Discussion by devlzcode** [tailwindlabs/tailwindcss#15409] [Help] [answered]
I want to add some custom variants but the [docs](https://tailwindcss.com/docs/plugins#parent-and-sibling-states) specify that `group-*` and `peer-*` don't work out of the box with custom variants and I should add variants for those specifically. However in my findings.. they do? Also using `addVariant("group-<variantName>", ...)` doesn't account for named groups either so whats the tailwind way of doing that?

**Comment by wongjn** [comment]
You can try following the core plugin code for `group-*` and `peer-*`:

https://github.com/tailwindlabs/tailwindcss/blob/f875ab9706cae8262e15e5b382580fc8e2d4197f/src/corePlugins.js#L155-L200

With some slight variation to work in custom land: https://play.tailwindcss.com/bQynqFD6AM

**Reply by devlzcode** [reply]
Yea I saw your usage of [`matchVariant`](https://github.com/tailwindlabs/tailwindcss/discussions/13464#discussioncomment-9061837) however I'm still confused on why the docs specify matching '&' over the default behavior of '*'. Is there really any reason to create `matchVariant("group-foo", ...)` over `addVariant("foo")` and letting tailwind handle the group/peer clauses?

**Reply by wongjn** [reply]
Could you give an example where `addVariant()` automatically registers relevant `group-*`/`peer-*` variants?

**Reply by devlzcode** [reply]
> Could you give an example where `addVariant()` automatically registers relevant `group-*`/`peer-*` variants?

Using Tailwind CSS v4.0 Beta:
```html
   <div class="group" data-test>
      <div class="group-test:text-red-500">
        Hello, world!
      </div>
    </div>
```
```css
@import "tailwindcss";
@plugin "./plugin.js";
```
```js
import createPlugin from "tailwindcss/plugin";

export default createPlugin(({ addVariant }) => {
    addVariant("test", "&[data-test]")
})
```

**Reply by wongjn** [reply]
Ah OK, v4 is a different story altogether!

**Reply by adamwathan** [reply, marked as answer]
In v4 the group and peer stuff is indeed all handled automatically 👍  Highly recommend using the CSS syntax by the way instead of the JS APIs, using the JS APIs triggers some very slow backward compatibility code paths that are otherwise completely skipped if we see you are only using modern APIs.

```css
@variant test (&[data-test]);
```

**Reply by devlzcode** [reply]
How would I write multiple selectors for a variant or override a previous variant? Afaik this doesn't work (just tried):
```css
@variant invalid (&:where([data-rac])[data-invalid]);
@variant invalid (&:where(:not([data-rac])):invalid);
```

**Reply by devlzcode** [reply]
Figured it out, its either:
```css
@variant invalid (&:where([data-rac])[data-invalid], &:where(:not([data-rac])):invalid);
```
or
```css
@variant invalid {
  &:where([data-rac])[data-invalid] {
    @slot;
  }

  &:where(:not([data-rac])):invalid {
    @slot;
  }
}
```
Though I'm not sure which one is better for this specific use case.

**Reply by devlzcode** [reply]
And what about variant values 😅 ?

**Reply by asieverding** [reply]
> > Could you give an example where `addVariant()` automatically registers relevant `group-*`/`peer-*` variants?
> 
> Using Tailwind CSS v4.0 Beta:
> 
> ```
>    <div class="group" data-test>
>       <div class="group-test:text-red-500">
>         Hello, world!
>       </div>
>     </div>
> ```
> 
> ```css
> @import "tailwindcss";
> @plugin "./plugin.js";
> ```
> 
> ```js
> import createPlugin from "tailwindcss/plugin";
> 
> export default createPlugin(({ addVariant }) => {
>     addVariant("test", "&[data-test]")
> })
> ```

How to accomplish this with v3?
Want to give my parent element the class `group/item` and another class set/unset by js `opened`.
So I want to use classes like `hidden group-opened/item:block` in my child elements.
